### PR TITLE
Implement Lessons and Questions

### PR DESCRIPTION
**Pull request by sherrpass**
### Description

1. Implemented Lessons
- Lesson Selection, Overview, Vocab and Quiz view controllers and view models
2. Implemented Questions
- Question, Question Formats, Question Types Structure
  - Question types can share the same question format. Question format is coupled with the rendering of the questions.
- QuestionsGenerator that takes in a scope of vocabs and QuestionGeneratorSettings and generates a QuestionSequence.
- QuestionSequence, a sequence of questions, generates a Question with every next() call.
- Question Constructors that constructs Question, one for each Question Type.
- QuestionScopeFactory, generated by QuestionsGenerator used by QuestionSequence to generate the scope of vocabs to create the next Question.
- QuestionConstructorRandomFactory, generated by QuestionsGenerator, used by QuestionSequence to generate the next QuestionConstructor randomly to create the next Question.
- A User that wishes to generate questions will use QuestionsGenerator to generate a QuestionSequence, customised through the use of QuestionGeneratorSettings. User will call QuestionSequence.next() to get the next question. QuestionSequence does not persist generated Questions. User is responsible for doing so if needed.
  - Rationale for returning a QuestionSequence instead of a collection of Questions is because quizzes may not know the exact finite number of Questions needed. (eg the Lesson Quiz generates questions till the user runs out of lives or gets 10 questions correct.) QuestionSequence represents a lazy infinite sequence of questions that are generated based on the User's preferred settings. It also abstracts away the details of keeping state to ensure settings are met so users do not have to. (eg Ensuring test all vocabulary in the compulsory scope is tested at least once)

3. Integrated into App, Separated Storyboards to avoid clutter and ease merge conflicts

### Diagram:
[Link to Draw.io](https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=CS3217_LingoClash_Question_Diagram#R7V1bb9s6Ev41BtKHLETdbD3GdtNTbIKmTdBzdl8WjMzYamXRR6KbuL%2F%2BkBIpSyJ1SWxdnDUQtBFFMtLMN5wLh6ORMVu%2FfArhZnWLF8gf6driZWTMR7quTwxA%2F2Mtu6QF6MYkaVmG3oK37Rvuvd%2BIN2q8destUJTrSDD2ibfJN7o4CJBLcm0wDPFzvtsT9vN%2FdQOXSGq4d6Evt%2F7pLciKtwLb2d%2F4A3nLFf%2FTE32c3FhD0Zm%2FSbSCC%2FycaTI%2BjoxZiDFJflu%2FzJDPqCfokoy7LrmbPliIAtJkwPxyPJmtft6G5L%2Bff%2F0wHz%2Bj%2F325BFYyzS%2Fob%2Fkbf92iiHg4oK2fUIBCSHDI34DsBFmiZ2%2Ftw4BeTZ9wQK7h2vMZi%2F9A%2Fi9EPBfSG3xmFBL0UvrMIKUExRDCa0TCHe3CB0wsTjwBH41fP2d4Ydq8cZXhg2XyRsgBsEwn39OI%2FsLJ9AqSWYZEDbSgmOGXOCQrvMQB9D%2FuW6ch3gYLxKbV6NW%2Bzw3GG9oIaOMPRMiOCwDcEkybVmTt87uUYOHuLzb%2BX5a4%2FA%2BfLr6Yv%2BSudvyqlDsLGK3iBwIpq9hrlDKqEj4R3oYuqurIpZ7AcImqZmS9VIAIkQ%2BJ9yv%2FfEfnLRCo6Za5p8YkoNn9cgmcudSAS8b4QC7FQ6%2FCEO4yHTbYC0iUmfmONezXbKAVF23TyCumugG6rRUwkjzDHjHpyxwCIv0MoiaifuiC3BOIjHEXILImgwIRChZXzOilza4Po8hzk8Zrz%2FcHB63EE%2BgaWaZTCayWcDIpNbOjxka2ABXjIvS9ZcC4TJmGQtrAzGyKAf%2BK3yAMSlN35fmLG7jDW8aIiED3p7iarnDo%2FabTQoEMejskHHO6netxz0ZyvIYoon3uBF5AoekWvuQ63sCI8AYX%2Bz7cRN5j%2BhprihEvmGJC8LrOYG3kTlRAt9TJGOt5RDi25GMYmsLFmNht2TeOhJZLevk3R8yMgoaEW5dChv7lK3pnZE2%2FyjdH1lyCFKUdiVkd4p9ohn3ay5gHOMEYXSQKTQJmPnoipSCLNtD1guVN3Gdu7lu%2BcVqxJkzHPvnxyrTyFgsUxOsggQQmaGCs56JLH9Sa0h9K8hnzdthrzOg12F%2FTH9Y9JMnrQi%2FmP6JQe0YMbq2gqGS5So2WZqjR20KNQHEGNRL7fS9ma8J%2BEc8Ab%2BL9mnIx1mac2Q%2Bx7roEEiAMGRCGgvk%2BfET%2BHY68OPZgzMOkbwEUw%2BO7pTdcLdpiO5DYvkw0CkpVzEVEzQlK%2F%2BhDsl6IG%2FeIrikB1ebnZaIruNhyAKvjZUIvNUVSS%2BSew%2BVskfRkkWhqhziHGlOBGqC1tswYEmyC7foRhV%2Be0mUmWVw%2Bsze%2FHjEAG4DRT9gtdyF%2BhI%2BeT1d4xPvOPZfdgewBbZ9hQEz2sNsgJs%2B0D94y9uj2kt3PT%2B3i9WbrRzjcPbBhwfLexXRcPDcFsZj0O3bho3qGqMGA8%2Bp4NOCXhXcrjCgVzsdmWzA3z0ZUD3xX7esoGd%2FW8mbIxnNmHyzncdXoxG41hZFXFBOZjJZqd8zQ29odA5YlUaiDiFguetokPPboY2Y%2FZIJj2gHBMcGG2uAYDzZ1EBw7iIumrO1vIXFXsV58wHP05AXxGqMKOxTZ3yRw%2BXYBMTQ1ocQ0Tk5ADEsSEF25fdyWfBhyJHA000dXU2odEUztUXY1nWYslCYk7tEuN7V3ZJfXgMko7DGMDRlNKmcO6G3ZK4YcKhzpU9bAJDViVjFazOjqSvbG%2BR5aSVdMViiMJTvK9IyD5mfD92jIqg4LmLLiViLJaAtI5jl62AffhSNTx%2Fe27F5Tjh4mi4IrtI1QPxfZFSX2zNNlJLOAfKBDL%2FeOcxppjJ3u82rSEaomckxaiSqRAtGCF2CfvYAKqbNOwwuQk0NFgIxKVSDEnFquAeWc60Em6V82pX5BMpQ%2BSnb0oN0FABR6uVN%2FAVjjsyRVQNTuSpLelGxigEIgX7eyiKjtDwyrgKAWklMEDWvEfO%2F0vysR1xWhx45F3OxFWaYZ5elFnE%2B%2BTy9%2FbUb5W4W9eWaZpjeUdpD4xZ1nLZqFeGwhCbF%2BgM7PlbSc%2BtpL1uK7Bpx2ooADFqgZUFRJ%2BQEtIVQQXrEfU5q0ck5O6GTLCYACIgxF8EqZ0yJSpY6vQTWVp8LiDwF6IRfnoEQ3mU9lRwWGkiEJTLMPvTeUIx%2FasH0lYBaPfBg5W6h%2BgFBVZQPMSWX%2FtlSZ7F4pXKdvMFjg9TVkv%2B%2FOqm0gqg0AeYeva9Um7%2FBR6VUgqFTNzXAUlG0cn9XdAequ5BDscNRdL6k4Q1F3xnjg6s4suP58z6Rc3RUGCFeusbozulB3guwqz43l2p5V3LBUXFqfpG6dGrem4hRJLFRiY7RcPDGlpTr69iGzAZ3Vehck2aEuSelmGeXxfnWUdC9P%2Bk7%2B%2F3BWmUfEZllFgleqTLMtldnPtvUhAU5RCeNoiaeH1ow4jP69bHZmaus4jpMPThsTsyY8HV%2FdodCjJGDa6eCYdWOWNs0i6DWJwO6lXFIn9G8aT%2Bm1WM6k370ebdTZXk9TvonSbVm%2B%2Fbi%2Fsyeb6NPXl9t%2FQ%2F9P3yTTvy5Lwpkt%2BwWF%2BhQTUJ0yUOyvd2HlT5wzqApqU2u4GJi9gGqcB4mj1YCq0F%2BcF2oVVI58NOIboogJqR0swa2jAoSF0iC2IztLE0dloQqz9fj6VD6hmDmpxh1eXfvuoedRfHKN%2Bim%2Bj%2BTQX7vn1pwSo4NPYxW2oVX578qDa0ZrdJW3EB%2Be8dyLflDRoF5iPunp%2B0yiZ355e155BN1TJ47dfQ7hJr%2BUvZ2wdskCUgJY%2BbwTcCyZrnZrZJV3P%2B699cZHp0nP9HhjbwQdy0fzBBET4T8VUupm76RUHZ4qmjVyimsujp3Pd6VX3HoBep6qeRZUZNS%2BLUeqgSXzKpZleSLo%2FzoTRbIpdBG6F0dfrPwMiYnFB1XYJsV5dKswUWKCSRO91mZK8Zl%2F3tLHMtWP1a4d3uDU1kAQPK6DMHrxSDqO%2Fp4M0y1%2BuR%2FFLo6D%2BybOYonz3oGAgOLC92YJKYpaSxICapIPJRHpokSnXV56UWEv8wr2QzKVQXEbT3WEXaE226vxYMt6s8pUvu3NHqkmbNFUVpSW79YckRdzpal8IvQEVu8E1UtFn5vKJ0JKlXvcMSnlXf2hGhqtmcpN6xonsOvFZHDeaDEY4%2Bp5WjIYnAHaC2NFgZaBIr3WpH7PUE9D1Idax0bxYN9QrONJFzF3cdD2ZHZnK7cO6t25pnvm9qHnuQ5St86wPnDSPluabqWPS872dvThGe3%2FTFzGjbOGShRBR%2BIiO6QtO%2FLVu4aXxWMEijPmE1V5y%2Fa%2BWla%2Ba0Bbr3G4hqTr0pZOyRJbktigcCd1FQ3TGgqH0LA8H6RJ6b44nfQxzBHT%2FnvLvvQ33cDFgm1ox%2Fmp2uaF%2FhtTUUvSgzPtYoRIRa2IsUice4MbW19XoYTVpRw0ahmYxrY6cWIVtX275J8qmHMMxr2pRsZrGAdU4YdOOdez5AluJevkSfBMVJrpj2eVXxZl5d0Hpm8sO%2B9%2FqfSNriCgeQx9U56K%2FYoivIcBswWK1xxFydNbGX%2FvErBy4KeqstmJEbsYnO99SZfzgsoKTJ0YpZssxKrUttYoLTtH%2BY9s9KPNXgXXdC%2BzP7i%2BE89fJG7Xev4jyfFvL0e%2BwUcam8Voj5Y08E7ib0253TTef2jy%2BmEiqPfBlMFULWvMzKax1K5Cdsp1Q1F3ug%2BP7hjarzHnmkdOiqUBFGVogcoXaU39AbOXEz%2FNytA2kKimx3iUFf%2FKAXyo4nv1nmLx9L5ds6noVHZvqQSEdXLaszE8RFBgqN9N17UCx8dWJxzvRTUfrdLM8dmv98J9YDvjHPetmhIxdQPagsv7OD6aSQNyfRhFnjvK5v50Ab7hqCYJSvakBnvaId1F0dGWgSqn5N6gKGKxKe83SxlXbzMfP4RSbS4Cu%2FDB2LEqXnKkhFF6GWJmTO8pTd9plVDC%2BPgP)
CS3217_LingoClash_Question_Diagram drawio

### TODO
Link with Firebase backend to load actual data instead of hardcoded data.

### CODE REVIEW

File: LingoClash/LingoClash/Model/Questions/QuestionFormats/QuestionFormat.swift [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//
+//  QuestionFormat.swift
+//  LingoClash
+//
+//  Created by Sherwin Poh on 22/3/22.
+//
+
+/// For users of questions to know and access the types of question formats.
+enum QuestionFormat {
```
Comment (kevinchua6) [inline LingoClash/LingoClash/Model/Questions/QuestionFormats/QuestionFormat.swift:9]: Seems like this isn't used?

File: LingoClash/LingoClash/Model/Questions/QuestionsGenerator/QuestionsGenerator.swift [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            return nil
+        }
+    }
+}
+
+struct QuestionGeneratorSettings {
+    // fixed number of questions to generate, if nil, unlimited
+    let numberOfQuestions: Int?
+    // probability of questions types to generate, if nil, uniformly distributed
+    let questionProbabilities: Dictionary<QuestionType, Double>?
```
Comment (kevinchua6) [inline LingoClash/LingoClash/Model/Questions/QuestionsGenerator/QuestionsGenerator.swift:86]: I'm actually not quite sure how this works. Is it if the Double is bigger it has a higher chance of appearing?

File: LingoClash/LingoClash/Model/Currency/Transaction.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
+//
+//  Transaction.swift
+//  LingoClash
+//
+//  Created by Sherwin Poh on 19/3/22.
+//
+
+import Foundation
+
+class Transaction {
```
Comment (kevinchua6) [inline LingoClash/LingoClash/Model/Currency/Transaction.swift:10]: Should this be a protocol? Not sure what design you're going for haha

**Review by kevinchua6 — APPROVED** [review]
Looks good to me! We are also using Combine instead of Dynamic but that can be changed later when you are free ig

File: LingoClash/LingoClash/Model/Questions/QuestionsGenerator/QuestionsGenerator.swift (hunk 2 above)
Comment (sherrpass) [reply LingoClash/LingoClash/Model/Questions/QuestionsGenerator/QuestionsGenerator.swift:86]: yep

File: LingoClash/LingoClash/Model/Currency/Transaction.swift (hunk 3 above)
Comment (sherrpass) [reply LingoClash/LingoClash/Model/Currency/Transaction.swift:10]: Right this is the rough work left when I was thinking how to implement the currency system. Will remove it in another PR.

**Pull request merged**